feat: Make it so PFE-Clipboard can copy arbitrary content

wesruv opened the pull request:
Need docs update and tests still, also think we can add a disabled state if it can't find what it should be copying.

For this proof o concept, see the comments I made below, and then check out the demo page here:
https://deploy-preview-1795--patternfly-elements.netlify.app/elements/pfe-clipboard/demo/

### Related issues

- #1794

<!--
### Preview

Link(s) to demo page(s) where this element can be viewed:
- https://deploy-preview-1795--patternfly-elements.netlify.app/elements/pfe-clipboard/demo/


### What has changed and why

- 


### Testing instructions

- [ ] *Test case #1*
  1. Go here
  2. Do this
- [ ] *Test case #2*
  1. Go here
  2. Do this


#### Browser requirements

Your component should work in all of the following environments:

- [ ] Latest 2 versions of Edge
- [ ] Internet Explorer 11 (should be useable, not pixel perfect)
- [ ] Latest 2 versions of Firefox (one on Mac OS, one of Windows OS)
- [ ] Firefox 78 (or latest version for Red Hat Enterprise Linux distribution)
- [ ] Latest 2 versions of Chrome (one on Mac OS, one of Windows OS)
- [ ] Latest 2 versions of Safari
- [ ] Android mobile device (such as the Galaxy S9)
- [ ] Apple mobile device (such as the iPhone X)
- [ ] Apple tablet device (such as the iPhone Pro)


### Ready-for-merge Checklist

- [ ] Expected files: all files in this pull request are related to one request or issue (no stragglers or scope-creep).
- [ ] Tests have been updated to cover these changes.
- [ ] Browser testing passed.
- [ ] Changelog updated.
- [ ] Documentation (README.md, WHY.md, etc.) updated or added.
- [ ] Link to the demo recording: []()
- [ ] Approved by designer.


### Merging

Please **squash** when merging and ensure your commit message uses [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) formatting.

**Be sure to share your updates with the [patternfly-elements-contribute@redhat.com](mailto:patternfly-elements-contribute@redhat.com) mailing list!**

-->

heyMP commented on elements/pfe-clipboard/src/pfe-clipboard.js:
```diff
@@ -51,6 +51,10 @@ class PfeClipboard extends PFElement {
+      target: {
+        type: String,
+        default: 'url',
+      }
```
I love this approach.  We could probably use this pattern for other elements. 👍

heyMP commented on elements/pfe-clipboard/demo/index.html:
```diff
@@ -63,11 +63,21 @@ <h2>Clipboard: with custom icon</h2>
+    </pfe-clipboard>
+    <script>
+      window.addEventListener('load', function() {
+        document.getElementById('propertyCopy').contentToCopy = '    <h2>Clipboard: with custom text & copying text from element</h2>\n    <pfe-clipboard role="button" tabindex="0" target="#textToCopy">\n      <span slot="text">This will copy the text in the text field below!</span>\n      <span slot="text--success">Making some copies!</span>\n    </pfe-clipboard>\n    <input type="text" id="text ...
```
ballin!

wesruv commented on elements/pfe-clipboard/src/pfe-clipboard.js:
```diff
@@ -162,34 +208,36 @@ class PfeClipboard extends PFElement {
    * @return {Promise<string>} url
    */
-  copyURLToClipboard() {
+  copyTextToClipboard(text) {
```
Sure!

wesruv commented:
The code is good to go as far as I'm concerned, but I it still needs docs. Wanted to bounce this issue off folks before I wrote the docs:

I added a few disabled states for when:
* `target='property'` and `el.contentToCopy` is not set
  * I added a getter and setter so I don't think there's a chance the component will be disabled if that value is set
* `target='selector'` and element in selector can't be found
  * This one *is not* fool proof. The code checks to see if the selector finds an element on `connectedCallback`, on `DOMContentLoaded`, and `load`, but if the content is injected dynamically the site would have to call `el.checkForCopyTarget()`
  * It also runs `checkForCopyTarget()` when the `target` attribute changes, which helps with the issue in the previous bullet
 
The only other ways of checking to see if the element exists would be heavy... like a MutationObserver on the document 🤢 ...

Do you all think this is good enough? Is there a 'cool kid' way to handle that 'not fool proof' concern that I'm missing?

wesruv commented:
> The code is good to go as far as I'm concerned, but I it still needs docs. Wanted to bounce this issue off folks before I wrote the docs:
>  ...
> Do you all think this is good enough? Is there a 'cool kid' way to handle that 'not fool proof' concern that I'm missing?

Discussed this with Kyle Buchanan, Michael Clayton, and other CPFED folks, and it feels like checking to see if an element exists isn't worth it. We can _know_ if the property is set, but if an element exists is too hairy, and not a necessary thing to guard against.

heyMP commented on elements/pfe-clipboard/src/pfe-clipboard.js:
```diff
@@ -91,12 +114,17 @@ class PfeClipboard extends PFElement {
     this.removeEventListener("click", this._clickHandler.bind(this));
     this.removeEventListener("keydown", this._keydownHandler.bind(this));
-    this.shadowRoot.removeEventListener("slotchange", this._slotchangeHandler.bind(this));
+    window.removeEventListener('DOMContentLoaded', this.checkForCopyTarget);
+    window.removeEventListener('load', this.checkForCopyTarget);
```
@wesruv I think we're removing these right?

wesruv replied:
D'OH!

heyMP commented:
@corey Here are the icon changes: 
https://deploy-preview-1795--patternfly-elements.netlify.app/elements/pfe-clipboard/demo/

- [x] Light theme Icon color: #6a6e73; Hover color: #151515
- [x] Dark theme Icon color: #d2d2d2; Hover color: #fff

I also added a saturated theme where I used the following variables.  I tried to semantically match them to their corresponding dark theme variables but they can be changed.
- icon color: `text--muted--on-saturated: #d2d2d2`
- icon color hover: text--on-saturated #fff`

coreyvickery commented:
@heyMP Looks great.

heyMP commented on elements/pfe-clipboard/src/pfe-clipboard.scss:
```diff
@@ -47,16 +53,43 @@ $LOCAL-VARIABLES: (
   }
 }
 
+// Attach variables to hover, focus states
+// ignore disabled.
+:host(:hover:not([disabled])),
+:host(:focus:not([disabled])) {
+  .pfe-clipboard {
+    &__text {
+      color: pfe-local(Color--hover) !important;
+    }
+    &__icon {
+      // Customize icon color for pfe-icons
+      --pfe-icon--Color: #{pfe-local(icon--Color--hover)};
+      svg {
+        // Customize icon color of fallback svg icon
+        fill: pfe-local(icon--Color--hover) !important;
+      }
+    }
+  }
+}
+
```
Simplified our old focus,hover selectors.

wesruv replied:
👍🏼

heyMP commented on elements/pfe-clipboard/src/pfe-clipboard.scss:
```diff
@@ -74,16 +107,14 @@ $LOCAL-VARIABLES: (
+  --pfe-clipboard--icon--Color: #{pfe-local(icon--Color--dark)};
+  --pfe-clipboard--icon--Color--hover: #{pfe-local(icon--Color--hover--dark)};
 }
 
-// Note: Hover states need to be defined after focus states
-:host(:not([aria-disabled="true"]):hover),
-:host(:not([aria-disabled="true"])) ::slotted(:hover) {
-  --pfe-clipboard--Color: #{pfe-local(Color--hover)};
+:host([on="saturated"]) {
+  --pfe-clipboard--icon--Color: #{pfe-local(icon--Color--saturated)};
+  --pfe-clipboard--icon--Color--hover: #{pfe-local(icon--Color--hover--saturated)};
```
Explicitly adding theme context hooks for icon color.  We don't have to include text color variables because that's automatically covered by broadcast variables.

wesruv commented:
Removed those unnecessary removeEventListeners

heyMP reviewed (approved):
LGTM!
